packages/react/src/SelectPanel/SelectPanel.tsx

```diff
@@ -87,19 +93,69 @@ export function SelectPanel({
   textInputProps,
   overlayProps,
   sx,
+  loading,
+  initialLoadingType = 'spinner',
   ...listProps
 }: SelectPanelProps): JSX.Element {
+  const inputRef = React.useRef<HTMLInputElement>(null)
   const titleId = useId()
   const subtitleId = useId()
+  const dataLoadedOnce = useRef(false)
+  const [isLoading, setIsLoading] = useProvidedStateOrCreate(loading, undefined, false)
   const [filterValue, setInternalFilterValue] = useProvidedStateOrCreate(externalFilterValue, undefined, '')
+  const {safeSetTimeout, safeClearTimeout} = useSafeTimeout()
+  const loadingDelayTimeoutId = useRef<number | null>(null)
   const onFilterChange: FilteredActionListProps['onFilterChange'] = useCallback(
     (value, e) => {
+      if (dataLoadedOnce.current) {
+        // If data has already been loaded once, delay the spinner a bit. This also helps
+        // not show and then immediately hide the spinner if items are loaded quickly, i.e.
+        // not async.
+
+        if (loadingDelayTimeoutId.current) {
+          safeClearTimeout(loadingDelayTimeoutId.current)
+        }
+
+        loadingDelayTimeoutId.current = safeSetTimeout(() => setIsLoading(true), 1000)
+      } else {
+        // If this is the first data load and there are no items, show the loading spinner
+        // immediately
+
+        if (items.length === 0) {
+          setIsLoading(true)
+        }
+      }
+
       externalOnFilterChange(value, e)
       setInternalFilterValue(value)
     },
-    [externalOnFilterChange, setInternalFilterValue],
+    [externalOnFilterChange, setInternalFilterValue, setIsLoading, safeSetTimeout, safeClearTimeout, items],
   )
 
+  useEffect(() => {
+    if (isLoading) {
+      setIsLoading(false)
+      dataLoadedOnce.current = true
+    }
+    // Only fire this effect if items have changed
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [items])
```

Comment:
Note for myself: Come back to see if this block can be refactored to remove the eslint warning while avoiding an accidental infinite loop

Reply:
Yeah, good call. I'm not 100% happy with this either, but it was the only approach I was able to get working. The idea here is that, if the component is in a loading state and items get updated, loading must be finished.